state/store.go

```diff
@@ -125,6 +125,102 @@ type ABCIResponses struct {
 	BeginBlock *abci.ResponseBeginBlock  `json:"begin_block"`
 }
 
+// PruneStates deletes states between the given heights (including from, excluding to). It is not
+// guaranteed to delete all states, since the last checkpointed state and states being pointed to by
+// e.g. `LastHeightChanged` must remain. The state at to must also exist.
+//
+// The from parameter is necessary since we can't do a key scan in a performant way due to the key
+// encoding not preserving ordering: https://github.com/tendermint/tendermint/issues/4567
+// This will cause some old states to be left behind when doing incremental partial prunes,
+// specifically older checkpoints and LastHeightChanged targets.
+func PruneStates(db dbm.DB, from int64, to int64) error {
```

Comment:
Should this be deleting txindex storage from these heights too?

Reply:
Hm, good question! I sort of considered that to be user data, but I'm not sure if it makes much sense to keep it when the blocks are gone. Can we even delete indexed data? The `TxIndexer` interface doesn't have any deletion methods.

Anyone else have an opinion here?

Reply:
1) If we were to use external indexer (Postgresql), would we also request removing these heights? I'd say no
2) how useful txindex is without blocks? one can query txs (without ability to prove them)

Reply:
Yeah, my take is that we consider indexed data to be "external data" that is simply exported as it was at the time. As a user, I think that's the behavior I'd be least surprised by, but I'm open to discussing this.